File: backend/routers/phase3.py
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session
from core.database import get_db
from core.security import get_current_user
from models.user import User
from models.agent import Agent
from models.organization import OrganizationMember
from schemas.agent_role import (
    AgentRoleCreate, AgentRoleUpdate, AgentRoleResponse,
    AgentSkillBindRequest, AgentSkillUpdate, AgentSkillResponse,
    AgentMemoryConfig, AgentMemoryResponse, AgentContextRequest,
    AgentHistoryResponse, AgentMetricsResponse, AgentDailyStatsResponse,
    AgentTaskCountResponse, OrgAgentUsageResponse, AgentPerformanceResponse,
    AgentHealthDetailResponse
)
from services.phase3_service import Phase3Service
from datetime import date

router = APIRouter(prefix="/api", tags=["phase3"])


def response(code: int = 0, message: str = "success", data=None):
    return {"code": code, "message": message, "data": data}


def get_org_id_from_user(db: Session, user: User) -> str | None:
    member = db.query(OrganizationMember).filter(
        OrganizationMember.user_id == user.id
    ).first()
    return member.org_id if member else None


def get_agent_or_404(db: Session, agent_id: str) -> Agent:
    agent = db.query(Agent).filter(Agent.id == agent_id).first()
    if not agent:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail={"code": 40401, "message": "Agent not found"}
        )
    return agent


def check_agent_access(db: Session, agent: Agent, user: User):
    org_id = get_org_id_from_user(db, user)
    if not org_id or agent.org_id != org_id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail={"code": 40301, "message": "Access denied"}
        )


# ========== Agent Roles (T-301 ~ T-308) ==========

@router.get("/agent-roles", response_model=dict)
def list_agent_roles(current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    """GET /api/agent-roles - List roles (T-301)"""
    org_id = get_org_id_from_user(db, current_user)
    if not org_id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail={"code": 40301, "message": "User does not belong to any organization"}
        )
    service = Phase3Service(db)
    roles = service.get_org_roles(org_id)
    return response(data=[AgentRoleResponse.model_validate(r).model_dump() for r in roles])


@router.post("/agent-roles", response_model=dict)
def create_agent_role(data: AgentRoleCreate, current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    """POST /api/agent-roles - Create role (T-302)"""
    service = Phase3Service(db)
    role = service.create_role(data, current_user)
    return response(data=AgentRoleResponse.model_validate(role).model_dump())


@router.get("/agent-roles/{role_id}", response_model=dict)
def get_agent_role(role_id: str, current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    """GET /api/agent-roles/{id} - Get role (T-303)"""
    service = Phase3Service(db)
    role = service.get_role_by_id(role_id)
    if not role:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail={"code": 40401, "message": "Role not found"}
        )

    org_id = get_org_id_from_user(db, current_user)
    if role.org_id != org_id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail={"code": 40301, "message": "Access denied"}
        )
    return response(data=AgentRoleResponse.model_validate(role).model_dump())


@router.put("/agent-roles/{role_id}", response_model=dict)
def update_agent_role(role_id: str, data: AgentRoleUpdate, current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    """PUT /api/agent-roles/{id} - Update role (T-304)"""
    service = Phase3Service(db)
    role = service.get_role_by_id(role_id)
    if not role:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail={"code": 40401, "message": "Role not found"}
        )

    org_id = get_org_id_from_user(db, current_user)
    if role.org_id != org_id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail={"code": 40301, "message": "Access denied"}
        )

    role = service.update_role(role, data)
    return response(data=AgentRoleResponse.model_validate(role).model_dump())


@router.delete("/agent-roles/{role_id}", response_model=dict)
def delete_agent_role(role_id: str, current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    """DELETE /api/agent-roles/{id} - Delete role (T-305)"""
    service = Phase3Service(db)
    role = service.get_role_by_id(role_id)
    if not role:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail={"code": 40401, "message": "Role not found"}
        )

    org_id = get_org_id_from_user(db, current_user)
    if role.org_id != org_id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail={"code": 40301, "message": "Access denied"}
        )

    service.delete_role(role)
    return response(message="Role deleted successfully")


@router.post("/agents/{agent_id}/skills", response_model=dict)
def bind_agent_skill(agent_id: str, data: AgentSkillBindRequest, current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    """POST /api/agents/{id}/skills - Bind skill (T-306)"""
    agent = get_agent_or_404(db, agent_id)
    check_agent_access(db, agent, current_user)

    service = Phase3Service(db)
    skill = service.bind_skill_to_agent(agent, data)
    return response(data=AgentSkillResponse.model_validate(skill).model_dump())


@router.delete("/agents/{agent_id}/skills/{skill_id}", response_model=dict)
def unbind_agent_skill(agent_id: str, skill_id: str, current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    """DELETE /api/agents/{id}/skills/{skill_id} - Unbind skill (T-307)"""
    agent = get_agent_or_404(db, agent_id)
    check_agent_access(db, agent, current_user)

    service = Phase3Service(db)
    skill = service.get_skill_by_id(skill_id)
    if not skill or skill.agent_id != agent_id:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail={"code": 40401, "message": "Skill not found"}
        )

    service.unbind_skill(skill)
    return response(message="Skill unbound successfully")


@router.get("/agents/{agent_id}/skills", response_model=dict)
def list_agent_skills(agent_id: str, current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    """GET /api/agents/{id}/skills - List agent skills (T-308)"""
    agent = get_agent_or_404(db, agent_id)
    check_agent_access(db, agent, current_user)

    service = Phase3Service(db)
    skills = service.get_agent_skills(agent_id)
    return response(data=[AgentSkillResponse.model_validate(s).model_dump() for s in skills])


# ========== Agent Memory (T-310 ~ T-315) ==========

@router.get("/agents/{agent_id}/memory", response_model=dict)
def get_agent_memory(agent_id: str, current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    """GET /api/agents/{id}/memory - Get memory config (T-310)"""
    agent = get_agent_or_404(db, agent_id)
    check_agent_access(db, agent, current_user)

    service = Phase3Service(db)
    memory = service.get_agent_memory_config(agent)
    return response(data=memory)


@router.put("/agents/{agent_id}/memory", response_model=dict)
def update_agent_memory(agent_id: str, data: AgentMemoryConfig, current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    """PUT /api/agents/{id}/memory - Update memory config (T-311)"""
    agent = get_agent_or_404(db, agent_id)
    check_agent_access(db, agent, current_user)

    service = Phase3Service(db)
    memory = service.update_agent_memory_config(agent, data)
    return response(data=memory)


@router.post("/agents/{agent_id}/context", response_model=dict)
def set_agent_context(agent_id: str, data: AgentContextRequest, current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    """POST /api/agents/{id}/context - Set context (T-312)"""
    agent = get_agent_or_404(db, agent_id)
    check_agent_access(db, agent, current_user)

    service = Phase3Service(db)
    result = service.set_agent_context(agent, data)
    return response(data=result)


@router.get("/agents/{agent_id}/history", response_model=dict)
def get_agent_history(agent_id: str, limit: int = Query(50, ge=1, le=200), current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    """GET /api/agents/{id}/history - Get conversation history (T-313)"""
    agent = get_agent_or_404(db, agent_id)
    check_agent_access(db, agent, current_user)

    service = Phase3Service(db)
    history = service.get_agent_history(agent, limit)
    return response(data=history)


@router.delete("/agents/{agent_id}/memory/clear", response_model=dict)
def clear_agent_memory(agent_id: str, current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    """DELETE /api/agents/{id}/memory/clear - Clear memory (T-314)"""
    agent = get_agent_or_404(db, agent_id)
    check_agent_access(db, agent, current_user)

    service = Phase3Service(db)
    result = service.clear_agent_memory(agent)
    return response(data=result)


@router.post("/agents/{agent_id}/reset", response_model=dict)
def reset_agent_state(agent_id: str, current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    """POST /api/agents/{id}/reset - Reset agent state (T-315)"""
    agent = get_agent_or_404(db, agent_id)
    check_agent_access(db, agent, current_user)

    service = Phase3Service(db)
    result = service.reset_agent_state(agent)
    return response(data=result)


# ========== Agent Metrics (T-320 ~ T-325) ==========

@router.get("/agents/{agent_id}/metrics", response_model=dict)
def get_agent_metrics(agent_id: str, days: int = Query(7, ge=1, le=90), current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    """GET /api/agents/{id}/metrics - Get metrics (T-320)"""
    agent = get_agent_or_404(db, agent_id)
    check_agent_access(db, agent, current_user)

    service = Phase3Service(db)
    metrics = service.get_agent_metrics(agent_id, days)
    return response(data=[AgentMetricsResponse.model_validate(m).model_dump() for m in metrics])


@router.get("/agents/{agent_id}/metrics/daily", response_model=dict)
def get_agent_daily_stats(agent_id: str, start_date: date = Query(...), end_date: date = Query(...), current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    """GET /api/agents/{id}/metrics/daily - Get daily stats (T-321)"""
    agent = get_agent_or_404(db, agent_id)
    check_agent_access(db, agent, current_user)

    service = Phase3Service(db)
    metrics = service.get_agent_daily_stats(agent_id, start_date, end_date)
    return response(data=[AgentDailyStatsResponse.model_validate(m).model_dump() for m in metrics])


@router.get("/agents/{agent_id}/tasks/count", response_model=dict)
def get_agent_task_counts(agent_id: str, current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    """GET /api/agents/{id}/tasks/count - Get task counts (T-322)"""
    agent = get_agent_or_404(db, agent_id)
    check_agent_access(db, agent, current_user)

    service = Phase3Service(db)
    counts = service.get_agent_task_counts(agent_id)
    return response(data=counts)


@router.get("/orgs/{org_id}/agents/usage", response_model=dict)
def get_org_agent_usage(org_id: str, current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    """GET /api/orgs/{id}/agents/usage - Get org agent usage (T-323)"""
    org_id_from_user = get_org_id_from_user(db, current_user)
    if org_id != org_id_from_user:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail={"code": 40301, "message": "Access denied"}
        )

    service = Phase3Service(db)
    usage = service.get_org_agent_usage(org_id)
    return response(data=usage)


@router.get("/agents/{agent_id}/performance", response_model=dict)
def get_agent_performance(agent_id: str, days: int = Query(7, ge=1, le=90), current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    """GET /api/agents/{id}/performance - Get performance report (T-324)"""
    agent = get_agent_or_404(db, agent_id)
    check_agent_access(db, agent, current_user)

    service = Phase3Service(db)
    perf = service.get_agent_performance(agent, days)
    return response(data=perf)


@router.get("/agents/{agent_id}/health", response_model=dict)
def get_agent_health_detail(agent_id: str, current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    """GET /api/agents/{id}/health - Get agent health (T-325)"""
    agent = get_agent_or_404(db, agent_id)
    check_agent_access(db, agent, current_user)

    service = Phase3Service(db)
    health = service.get_agent_health_detail(agent)
    return response(data=health)

File: frontend/src/api/phase3.ts
import api from './index'
import type { ApiResponse } from '@/types'

// ========== Agent Roles ==========

export interface AgentRole {
  id: string
  name: string
  description: string | null
  system_prompt_template: string | null
  default_config: Record<string, unknown> | null
  org_id: string
  created_at: string
  updated_at: string
}

export interface AgentRoleCreate {
  name: string
  description?: string
  system_prompt_template?: string
  default_config?: Record<string, unknown>
  org_id?: string
}

export interface AgentRoleUpdate {
  name?: string
  description?: string
  system_prompt_template?: string
  default_config?: Record<string, unknown>
}

export const listAgentRoles = () =>
  api.get<ApiResponse<AgentRole[]>>('/agent-roles')

export const createAgentRole = (data: AgentRoleCreate) =>
  api.post<ApiResponse<AgentRole>>('/agent-roles', data)

export const getAgentRole = (id: string) =>
  api.get<ApiResponse<AgentRole>>(`/agent-roles/${id}`)

export const updateAgentRole = (id: string, data: AgentRoleUpdate) =>
  api.put<ApiResponse<AgentRole>>(`/agent-roles/${id}`, data)

export const deleteAgentRole = (id: string) =>
  api.delete<ApiResponse>(`/agent-roles/${id}`)

// ========== Agent Skills ==========

export interface AgentSkill {
  id: string
  agent_id: string
  skill_name: string
  skill_config: Record<string, unknown> | null
  enabled: boolean
  created_at: string
}

export interface AgentSkillBind {
  skill_name: string
  skill_config?: Record<string, unknown>
  enabled?: boolean
}

export const bindAgentSkill = (agentId: string, data: AgentSkillBind) =>
  api.post<ApiResponse<AgentSkill>>(`/agents/${agentId}/skills`, data)

export const listAgentSkills = (agentId: string) =>
  api.get<ApiResponse<AgentSkill[]>>(`/agents/${agentId}/skills`)

export const unbindAgentSkill = (agentId: string, skillId: string) =>
  api.delete<ApiResponse>(`/agents/${agentId}/skills/${skillId}`)

export const updateAgentSkill = (agentId: string, skillId: string, data: { enabled?: boolean; skill_config?: Record<string, unknown> }) =>
  api.put<ApiResponse<AgentSkill>>(`/agents/${agentId}/skills/${skillId}`, data)

// ========== Agent Memory ==========

export interface AgentMemoryConfig {
  agent_id: string
  memory_type: string
  max_context_tokens: number
  context_window: number
  persist_context: boolean
  context_items: number
}

export interface AgentContextRequest {
  context: Record<string, unknown>
}

export interface AgentHistory {
  agent_id: string
  messages: Record<string, unknown>[]
  total: number
}

export const getAgentMemory = (agentId: string) =>
  api.get<ApiResponse<AgentMemoryConfig>>(`/agents/${agentId}/memory`)

export const updateAgentMemory = (agentId: string, data: AgentMemoryConfig) =>
  api.put<ApiResponse<AgentMemoryConfig>>(`/agents/${agentId}/memory`, data)

export const setAgentContext = (agentId: string, data: AgentContextRequest) =>
  api.post<ApiResponse<{ agent_id: string; message: string }>>(`/agents/${agentId}/context`, data)

export const getAgentHistory = (agentId: string, limit = 50) =>
  api.get<ApiResponse<AgentHistory>>(`/agents/${agentId}/history?limit=${limit}`)

export const clearAgentMemory = (agentId: string) =>
  api.delete<ApiResponse<{ agent_id: string; message: string }>>(`/agents/${agentId}/memory/clear`)

export const resetAgentState = (agentId: string) =>
  api.post<ApiResponse<{ agent_id: string; status: string; message: string }>>(`/agents/${agentId}/reset`)

// ========== Agent Metrics ==========

export interface AgentMetric {
  id: string
  agent_id: string
  date: string
  tasks_completed: number
  tasks_failed: number
  avg_response_time_ms: number
  token_usage: number
  created_at: string
}

export interface AgentDailyStats {
  agent_id: string
  date: string
  tasks_completed: number
  tasks_failed: number
  avg_response_time_ms: number
  token_usage: number
}

export interface AgentTaskCount {
  agent_id: string
  total_tasks: number
  completed_tasks: number
  failed_tasks: number
}

export interface OrgAgentUsage {
  org_id: string
  total_agents: number
  active_agents: number
  total_tasks_completed: number
  total_token_usage: number
  agents: { id: string; name: string; status: string }[]
}

export interface AgentPerformance {
  agent_id: string
  period_start: string
  period_end: string
  total_tasks: number
  success_rate: number
  avg_response_time_ms: number
  avg_tokens_per_task: number
  uptime_percent: number
}

export interface AgentHealthDetail {
  id: string
  healthy: boolean
  status: string
  cpu_percent: number | null
  memory_mb: number | null
  uptime_seconds: number | null
  last_check_at: string
  error_count_today: number
}

export const getAgentMetrics = (agentId: string, days = 7) =>
  api.get<ApiResponse<AgentMetric[]>>(`/agents/${agentId}/metrics?days=${days}`)

export const getAgentDailyStats = (agentId: string, startDate: string, endDate: string) =>
  api.get<ApiResponse<AgentDailyStats[]>>(`/agents/${agentId}/metrics/daily?start_date=${startDate}&end_date=${endDate}`)

export const getAgentTaskCounts = (agentId: string) =>
  api.get<ApiResponse<AgentTaskCount>>(`/agents/${agentId}/tasks/count`)

export const getOrgAgentUsage = (orgId: string) =>
  api.get<ApiResponse<OrgAgentUsage>>(`/orgs/${orgId}/agents/usage`)

export const getAgentPerformance = (agentId: string, days = 7) =>
  api.get<ApiResponse<AgentPerformance>>(`/agents/${agentId}/performance?days=${days}`)

export const getAgentHealthDetail = (agentId: string) =>
  api.get<ApiResponse<AgentHealthDetail>>(`/agents/${agentId}/health`)